WWCodeTokyo/speak-her-db
# [Part 2] Migrate Topics API to Netlify functions

### References / ticket
- Part 2 of issue #15 
- Follow-up from PR https://github.com/WWCodeTokyo/speak-her-db/pull/143 **(already merged!)**
- Also solves issue https://github.com/WWCodeTokyo/speak-her-db/issues/144

### Proposed Changes

- [x] Continuation of the REST API hosted on Netlify Lambda functions under `src/lambda/` ([details here](https://github.com/WWCodeTokyo/speak-her-db/pull/143))
    - [x] For this PR, I created the `topics.js` function/endpoint
- [x] Added `getTopics` to the API service. It can be used across the app like this:
```javascript
import api from '@/services/api'; 
api.getLocations(successCallback, errorCallback);
```
- [x] Migrate `Topic` requests across the app from the Airtable plugin to the new API service
    - [x] Fetch Topics on `components/TopicsInput.vue`
    - [x] Fetch Topics on `views/FindSpeaker.vue`

- [x] Moved the request for Topics from `components/CardFooter` to `views/FindSpeaker` and passing the `topicList` down as props to `SpeakerCard` > `CardFooter` to fix this [N+1 query issue](https://github.com/WWCodeTokyo/speak-her-db/issues/144)

### Testing Plan

- Open [this deploy preview](https://deploy-preview-145--speak-her-db.netlify.app/)
- When the site is loading, open the network inspector and check that there's only 1 `topics` request now instead of many (there were ~70 before)
- On the network inspector click on the single `topics` request
- Request headers now show the endpoint is `/.netlify/functions/topics` and there's no authorization key on the request headers 🔒 

Alternatively, run it locally:
- Pull this branch locally, run `yarn install` to install the new dependencies
- If your connection to Airtable is already setup locally as per README instructions, just run the app with `yarn serve`
- Open `localhost:8080` on your browser and follow the steps above
